### PR TITLE
Attack chain, initial setup. (pull *immediately* for *any* TM issues)

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do
This PR:

- Implements a new attack chain, including non-attack item interactions, unified signals and return values, and documentation. It is based on /tg/ but is evolved to compensate for the difference in intent modes and primary/secondary action behavior.
- Preserves most of the existing attack chain, moving it to a set of "legacy" procs, and marking all objects using the legacy chain as such.
- Renames legacy attack chain procs in an obnoxious way to make it clear they don't belong and should be refactored whenever possible.
- Performs several migrations as demonstrations: cult daggers, `/obj/item/attackby`, and airlock controllers.
- Enforces a top level requirement that tool interactions must be on help intent. This probably changes some random obscure behavior somewhere but our intent system is so scuffed I doubt anyone will notice.
- Provides extensive documentation on how to perform attack chain migrations in the devdocs, complete with a full diagram of the attack chain.

This is not a wholesale migration for a good reason: just moving the existing procs to their legacy equivalents changed nearly 700 files. Attempting to perform this migration in one PR, moving every single item that utilized the attack chain over at once, would be a nightmare for me and reviewers and impossible to test thoroughly. This approach allows us to use both systems and gradually migrate over, with smaller PRs that are easier to write, review, and test. Getting the burndown list is as simple as searching the codebase for `__legacy__attackchain`. By providing documentation on how to perform the migration, we're enabling people to swarm on this so that it's not blocked by a single person.

I haven't touched `attack_hand()` yet.

## Why It's Good For The Game

Our current attack chain is a disorganized unruly mess. Procs are terribly named, signals are sent at the wrong times, nothing ever checks their parent proc to see if they should be running. If we want to enforce consistency and sanity when it comes to player interactions, this is the first step. If we want to make more uses of components and signals and be able to do so reliably, this is also critical.

## Images of changes

attack_chain_flowchart

## Testing
For basic attack chain redirection:

1. Verified placing items on table.
2. Verified placing things in food processor.
3. Verified attacking mobs with weapons.
4. Verified attacking myself with weapons.
5. Verified attacking vendor, and ensuring vendor tipping procced properly.
6. Spawned in as borgs with different modules, ensured tools and attacks worked.
7. Ensured energy gun fired when clicking from a distance.
8. Verified lighting cigarettes with welder.
9. Verified inserting screwdriver into autolathe with harm intent.
10. Verified attacking crematorium with welder on harm intent.
11. Verified injecting self with syringe.
12. Gave self telekinesis, verified moving things around and attacking using item.

Walked through all the above in the debugger and ensured expected procs were called.

For storage and duct tape migrations:

1. Spawned backpacks and duct tape.
2. Ensured duct tape still worked when used on non-storage objects.
3. Ensured duct tape was placed into backpack when used on it.

For cult dagger migration:

1. Ensured that cultists attacking non-cultists performed and logged attack.
2. Ensured that non-cultists using cult dagger were properly disarmed and hurt.
3. Ensured that cultists attacking cultists did not perform an attack.

For airlock controller migration:

1. Ensured that using item in-hand, either with mouse or keyboard shortcut,
   opened access UI.

<hr>

### Declaration

- [X] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/docs/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
<hr>

## Changelog

NPFC

### CODE REVIEW

File: code/_onclick/item_attack.dm [hunk 1]
```diff
@@ -1,155 +1,242 @@
...
+ *
+ * See: [/obj/item/proc/melee_attack_chain]
+ */
+/obj/item/proc/pre_attack(atom/A, mob/living/user, params)
+	SHOULD_CALL_PARENT(TRUE)
+
+	if(SEND_SIGNAL(src, COMSIG_PRE_ATTACK, A, user, params) & COMPONENT_CANCEL_ATTACK_CHAIN)
 		return TRUE
 
+	// TODO: Turn this into a component and have a sane implementation instead of extra-specific behavior in a core proc
```
Comment (BiancaWilkson) [inline code/_onclick/item_attack.dm:79]: Here's a TODO, plan on changing it in this PR or during the next steps?

File: code/game/atom/atom_tool_acts.dm [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ * or scanning someone with a health analyzer
+ */
+/atom/proc/base_item_interaction(mob/living/user, obj/item/tool, list/modifiers)
+	SHOULD_CALL_PARENT(TRUE)
+	PROTECTED_PROC(TRUE)
+
+	// We do not have a combat mode or secondary actions like /tg/, so instead
+	// I'm unilaterally deciding it here: If you are not on harm intent, tool
+	// interactions are not attacks. Shit like the autolathe accepting
+	// screwdrivers on harm intent is unintuitive and needs to go away, and there
```
Comment (BiancaWilkson) [inline code/game/atom/atom_tool_acts.dm:15]: What do you plan on replacing these with? As un-intuitive as it is, being completely unable to recycle screwdrivers seems even less intuitive.

File: code/game/gamemodes/cult/blood_magic.dm [hunk 3]
```diff
@@ -410,18 +410,18 @@
 /obj/item/melee/blood_magic/customised_abstract_text(mob/living/carbon/owner)
 	return "<span class='warning'>[owner.p_their(TRUE)] [owner.l_hand == src ? "left hand" : "right hand"] is burning in blood-red fire.</span>"
 
-/obj/item/melee/blood_magic/attack_self(mob/living/user)
-	afterattack(user, user, TRUE)
+/obj/item/melee/blood_magic/attack_self__legacy__attackchain(mob/living/user)
+	attackby__legacy__attackchain(user, user, TRUE)
```
Comment (BiancaWilkson) [inline code/game/gamemodes/cult/blood_magic.dm:414]: Should this be attackself__legacy_attackchain?

File: code/game/objects/items.dm [hunk 4]
```diff
@@ -398,56 +398,6 @@ GLOBAL_DATUM_INIT(welding_sparks, /mutable_appearance, mutable_appearance('icons
 			R.activate_module(src)
 			R.hud_used.update_robot_modules_display()
 
-// Due to storage type consolidation this should get used more now.
-// I have cleaned it up a little, but it could probably use more.  -Sayu
-/obj/item/attackby(obj/item/I, mob/user, params)
-	if(isstorage(I))
-		var/obj/item/storage/S = I
-		if(S.use_to_pickup)
```
Comment (BiancaWilkson) [inline code/game/objects/items.dm:406]: Just to make sure, did you test to make sure bags like chem/ore bags properly picked up stacks of the items they can carry?

**Review by BiancaWilkson — COMMENTED** [review]
Alright, I looked at the first 200 files and these are the questions/inconsistencies I have. More to come as I keep going down the list.

File: code/game/gamemodes/cult/blood_magic.dm (hunk 3 above)
Comment (BiancaWilkson) [reply code/game/gamemodes/cult/blood_magic.dm:414]: I tested it, and not only does it work fine but in hindsight I don't know what I was thinking when I said a proc should call itself.

File: code/game/objects/items.dm (hunk 4 above)
Comment (BiancaWilkson) [reply code/game/objects/items.dm:406]: Also tested this, it works properly

File: code/game/gamemodes/cult/blood_magic.dm (hunk 3 above)
Comment (warriorstar-orion) [reply code/game/gamemodes/cult/blood_magic.dm:414]: Thanks for taking the time to test it. I had to drop this and circle back to other things but I was absolutely going to.

File: code/_onclick/item_attack.dm (hunk 1 above)
Comment (warriorstar-orion) [reply code/_onclick/item_attack.dm:79]: Next steps. I don't want to risk having to test more things in the initial change.

**Review by BiancaWilkson — CHANGES_REQUESTED** [review]
A few days late but I finished looking at all the files. All the individual changes look great. I did manage to find a few bugs: Cult daggers can no longer properly unanchor cult girders, pylons, forges, altars, and archives. Giving those a tool_act should fix that if I'm understanding the refactor correctly. Additionally, there doesn't seem to be an attack animation or cooldown on hitting objects with other objects.

**Comment by warriorstar-orion** [comment]
I've re-made the attack chain diagram in draw.io: https://github.com/jgraph/drawio-desktop/releases/ and provided the source diagram in the docs directory. This means people don't have to resort to Lucidchart to modify it (since the number of objects on the chart is higher than allowed by Lucidchart's free tier), and makes it easier for people to update it if/when necessary.

**Review by BiancaWilkson — APPROVED** [review]
Bugs have been fixed, and after messing around with each of the majorly changed items I didn't notice any bugs. LGTM!

**Comment by SteelSlayer** [comment]
Testmerged this briefly, and we encountered a lot of issues. image

**Comment by warriorstar-orion** [comment]
> Testmerged this briefly, and we encountered a lot of issues. image


https://github.com/user-attachments/assets/4c212df9-80d4-44d7-85c3-e2dcfb4708bf

**Comment by warriorstar-orion** [comment]
Not ready for another TM just yet, will ping when it is.

**Review by Burzah — APPROVED** [review]
After stress testing this for 2 hours during a TM, this looks good to go.
This is an awesome foundation moving forward. 
Well done!

**Pull request merged**